sycl/doc/CompileTimeProperties.md

```diff
@@ -0,0 +1,680 @@
+corresponding LLVM IR.  One of the goals of this design is to avoid changes to
+the front-end each time we add a new property, so the front-end is not required
+to understand each property it consumes.  Instead, it follows a mechanical
+process for converting properties listed in the C++ annotation into LLVM IR,
+and this mechanical process need not be updated when we add new properties.
+
+Once the information about properties is represented in IR, it is available to
+compiler passes.  For example, the `sycl-post-link` tool might use a property
+in order to perform one of its transformations.  Some properties are consumed
+by the DPC++ compiler, but others are transformed into SPIR-V for use by the
+JIT compiler.  This design document also describes how this SPIR-V
+transformation is done.
+
+
+## Properties on a global variable type
+
+One use for compile-time properties is with types that are used exclusively
+for declaring global variables.  One such example is the
+[SYCL\_EXT\_ONEAPI\_DEVICE\_GLOBAL][2] extension:
+
+[2]: <extensions/DeviceGlobal/SYCL_INTEL_device_global.asciidoc>
+
+```
+namespace sycl::ext::oneapi {
+
+template <typename T, typename PropertyListT = property_list<>>
+class device_global {/*...*/};
+
+} // namespace sycl::ext::oneapi
+```
+
+The following code illustrates a `device_global` variable that is declared with
+two compile-time properties:
+
+```
+using sycl::ext::oneapi;
+
+device_global<int,
+  property_list_t<
+    device_image_scope::value_t,
+    host_access::value_t<host_access::access::read>>>
+  dm1;
+```
+
+The header file represents these properties with an internal C++ attribute
+named `[[__sycl_detail__::add_ir_global_variable_attributes()]]` whose value
+is a list that is created through a template parameter pack expansion:
+
+```
+namespace sycl::ext::oneapi {
+
+template <typename T, typename PropertyListT = property_list<>>
+class device_global {/*...*/};
+
+// Partial specialization to make PropertyListT visible as a parameter pack
+// of properties.
+template <typename T, typename ...Props>
+class
+#ifdef __SYCL_DEVICE_ONLY__
+  [[__sycl_detail__::add_ir_global_variable_attributes(
+    Props::meta_name..., Props::meta_value...
+    )]]
+#endif
+  device_global<T, property_list<Props...>> {/*...*/};
+
+} // namespace sycl::ext::oneapi
+```
+
+The initial entries in the C++ attribute's parameter list are the names of the
+properties, and these are followed by the values of the properties.  To
+illustrate using the same example as before, the result of the parameter pack
+expansion would look like this:
+
+```
+namespace sycl::ext::oneapi {
+
+template </* ... */> class
+#ifdef __SYCL_DEVICE_ONLY__
+  [[__sycl_detail__::add_ir_global_variable_attributes(
+    "sycl-device-image-scope",  // Name of first property
+    "sycl-host-access",         // Name of second property
+    "",                         // First property has no parameter
```

Comment:
This is a bit awkward -- there's no way to know how many properties there are without doing a linear scan of (potentially) the entire argument list to find an empty string argument and hope that signals the end of the property name list rather than a typo on the user's part. We'll have to be careful about diagnostics in this area to give users a reasonable shot at being able to use this attribute. (Also, is there a notion of default property values? e.g., do we have to worry about there legitimately being more names than values?)

Reply:
> there's no way to know how many properties there are without doing a linear scan of (potentially) the entire argument list to find an empty string argument and hope that signals the end of the property name list rather than a typo on the user's part

The list of property names is **not** terminated by an empty string, so the scan you propose is not the correct way to determine the number of properties.  Each property has exactly one value, so each property corresponds to two arguments in the attribute.  Assuming the optional braced "initializer list" argument is not specified, this means that the number of properties is *N/2* where *N* is the number of arguments.

I agree that the usage of these attributes is awkward for a human.  However, they are not intended for human use.  Rather, they are intended to be used as an implementation detail by our DPC++ headers.  (Note that these attributes are all in the `__sycl_detail__` namespace.)  Although the structure may look awkward for humans, it's actually quite natural for our expected usage in the headers.  For example, the reason all the property names come first in the argument list followed by all the property values is because this works particularly well when expanding the properties via a template parameter pack expansion.

> Some attributes already support expressions, but no attribute currently expects an init-list argument and so who knows how well this will work in practice.

If the "initializer list" is hard to implement, my fallback position is to accept a leading "count" argument, followed by a variable number of strings.  Usage would look like this:

```
[[__sycl_detail__::add_ir_kernel_parameter_attributes(

    // This is conceptually the same as {"sycl-no-alias", "sycl-foo"}.
    2, "sycl-no-alias", "sycl-foo",

    Props::meta_name..., Props::meta_value...
    )]]
```

Thus, the leading `2` says that the following 2 strings are the "filter list" to apply to the properties.

I thought that the braced list syntax was less awkward to use and less error prone.  However, if that's difficult to implement in the front-end, we can use this alternate syntax instead.

Reply:
> The list of property names is not terminated by an empty string, so the scan you propose is not the correct way to determine the number of properties. Each property has exactly one value, so each property corresponds to two arguments in the attribute. Assuming the optional braced "initializer list" argument is not specified, this means that the number of properties is N/2 where N is the number of arguments.

Ah, thank you for clarifying!

> Although the structure may look awkward for humans, it's actually quite natural for our expected usage in the headers.

Okay, good to know.

> I thought that the braced list syntax was less awkward to use and less error prone. However, if that's difficult to implement in the front-end, we can use this alternate syntax instead.

Good deal! I think either approach will work. I suspect the braced init list approach may be a bit more effort though. Because this is expected to be used only in your headers, is it reasonable to assume you'll have already included `<initializer_list>` so that's not really a concern either?

Reply:
> Because this is expected to be used only in your headers, is it reasonable to assume you'll have already included <initializer_list> so that's not really a concern either?

Yes, we can assume that the header includes `<initializer_list>`.  Just for my education, why is this necessary?  I was thinking that the braced list was just some C++ syntax.  It's not really an "initializer list", since we're not calling a constructor which takes `std::initializer_list`.  I'm not opposed to including `<initializer_list>` if that makes the front-end happier.  I'm just curious about the rationale.

Reply:
> Just for my education, why is this necessary?

Brace init lists are weird: https://godbolt.org/z/q5h1rsz7z and their use here is even weirder in some ways. Attributes are not things which are declared, so their argument lists are kind wonky because they have no declared type. So the question really boils down to "is that expression a list initialization of an object or not?". If it's a list initialization of an object, then http://eel.is/c++draft/dcl.dcl#dcl.init.list-2 applies. If it's not a list initialization of an object, then... we have to figure out what it means.

So it may not be necessary, strictly speaking. It's just breaking new ground.

Reply:
Thanks for the explanation, @AaronBallman.

Summarizing the outcome of this discussion, do you agree with the following?

* I'll keep the wording about the braced list for now, and we'll reevaluate if it turns out that implementation is too difficult.
* I'll update the design doc to state that the headers should include `<initializer_list>` when using the braced list syntax.

I understand that it may require some extra effort in the front-end to accept the braced list syntax.  I think it would be worth a little extra effort, but we shouldn't do this if it's a huge amount of work.  This seems like something that we can evaluate during implementation, and then adjust the design doc if necessary.

Reply:
I added the note about requiring `<initializer_list>` in 4ec288158b945ecf68f098dd82a53e7a263b3014.